File: lint-arwaky/src-python/taxonomy/governance_report_entity.py
from pydantic import BaseModel, ConfigDict, Field
from .lint_result_vo import LintResult, LintResultList
from .score_format_vo import Score
from .message_status_vo import ComplianceStatus, Count
from .lint_severity_vo import Severity
from .adapter_name_vo import AdapterName


class GovernanceReport(BaseModel):
    """Aggregated lint scan results entity."""

    model_config = ConfigDict(frozen=False)
    results: LintResultList = Field(default_factory=LintResultList)
    score: Score = Field(default_factory=lambda: Score(value=100.0))
    is_passing: ComplianceStatus = Field(
        default_factory=lambda: ComplianceStatus(value=True)
    )

    def add_result(self, result: LintResult):
        """Add a finding and update score."""
        # Since LintResultList is now a VO, we handle the mutation by creating a new
        # list if needed,
        # but for performance in entities, we can keep the internal list mutable if
        # encapsulated.
        # However, to be a true VO, the list should be replaced.
        new_values = list(self.results.values)
        new_values.append(result)
        self.results = LintResultList(values=new_values)
        self.score = self.score.deduct(result.severity)

    def update_compliance(self, threshold: Score):
        """Update compliance status based on score threshold and critical findings."""
        is_p = self.score.is_passing(threshold)
        has_critical = any(r.severity == Severity.CRITICAL for r in self.results)
        self.is_passing = ComplianceStatus(value=is_p and not has_critical)

    def results_by_source(self, source: AdapterName) -> LintResultList:
        """Filter results by adapter source."""
        source_str = str(source)
        return LintResultList(
            values=[r for r in self.results if str(r.source or "") == source_str]
        )

    @property
    def violation_count(self) -> Count:
        return Count(
            value=len([r for r in self.results if r.severity.score_impact > 0])
        )

File: lint-arwaky/src-rust/taxonomy/governance_report_entity.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct GovernanceReport {
    #[serde(default)]
    pub results: LintResultList,
    #[serde(default = "default_score")]
    pub score: Score,
    #[serde(default = "default_compliance")]
    pub is_passing: ComplianceStatus,
}

fn default_score() -> Score { Score::new(100.0) }
fn default_compliance() -> ComplianceStatus { ComplianceStatus::new(true) }

impl GovernanceReport {
    pub fn new() -> Self {
        Self { results: LintResultList::new(), score: Score::new(100.0), is_passing: ComplianceStatus::new(true) }
    }
    pub fn add_result(&mut self, result: LintResult) {
        self.results.push(result);
        self.score = self.score.deduct(result.severity);
    }
    pub fn update_compliance(&mut self, threshold: &Score) {
        let is_p = self.score.value >= threshold.value;
        let has_critical = self.results.values.iter().any(|r| r.severity == Severity::CRITICAL);
        self.is_passing = ComplianceStatus::new(is_p && !has_critical);
    }
    pub fn results_by_source(&self, source: &AdapterName) -> LintResultList {
        LintResultList {
            values: self.results.values.iter()
                .filter(|r| r.source.as_ref() == Some(source))
                .cloned()
                .collect(),
        }
    }
    pub fn violation_count(&self) -> Count {
        Count::new(self.results.values.iter()
            .filter(|r| r.severity.score_impact() > 0.0)
            .count() as i64)
    }
}

impl Default for GovernanceReport {
    fn default() -> Self { Self::new() }
}
